File: src/deadlock_build_sync/beam_display.py
"""Describe beam evidence and ability instructions."""

from __future__ import annotations

from dataclasses import replace
from typing import TYPE_CHECKING

from .build_support import numeric
from .value_validation import integer, object_dict, object_list, object_rows

if TYPE_CHECKING:
    from .purchase_types import PurchaseGuide

STATE_LABELS = {0: "Behind", 1: "Even", 2: "Ahead"}


def generator_metadata(guide: PurchaseGuide) -> dict[str, object]:
    return object_dict(guide.evidence_summary.get("generator")) or {}


def attach_beam_ability_names(
    guide: PurchaseGuide, kit: dict[str, object]
) -> PurchaseGuide:
    if not generator_metadata(guide):
        return guide
    return replace(
        guide,
        evidence_summary={
            **guide.evidence_summary,
            "ability_names": {
                str(row["id"]): str(row["name"])
                for row in object_rows(kit.get("abilities")) or []
            },
        },
    )


def _variant_state_rows(guide: PurchaseGuide) -> list[tuple[str, dict[str, object]]]:
    metadata = generator_metadata(guide)
    states = object_list(metadata.get("states")) or []
    evidence = object_dict(metadata.get("state_evidence")) or {}
    rows = []
    for state, label in STATE_LABELS.items():
        if state not in states:
            continue
        folds = object_dict(evidence.get(str(state))) or {}
        row = object_dict(folds.get("validation")) or {}
        count, wins = row.get("owners"), row.get("wins")
        if (
            type(count) is not int
            or type(wins) is not int
            or count <= 0
            or not 0 <= wins <= count
        ):
            continue
        rows.append((label, row))
    return rows


def variant_state_labels(guide: PurchaseGuide) -> str:
    return (
        ", ".join(label for label, _ in _variant_state_rows(guide)) or "State unknown"
    )


def variant_statistics(guide: PurchaseGuide, *, detailed: bool = False) -> list[str]:
    lines = []
    for label, row in _variant_state_rows(guide):
        count, wins = integer(row["owners"]), integer(row["wins"])
        text = f"{label}: {100 * wins / count:.1f}% | {wins}/{count} wins"
        if detailed:
            text += (
                f" | 95% interval {100 * numeric(row, 'lower_95'):.1f}%–{100 * numeric(row, 'upper_95'):.1f}%."
                f" State hero baseline: {100 * numeric(row, 'hero_win_rate'):.1f}% across {row['hero_matches']} matches."
            )
            cutoff = row.get("ownership_before_seconds")
            if type(cutoff) is int and cutoff > 0:
                text += f" All final core items owned before {cutoff} seconds; additional items allowed."
        lines.append(text)
    return lines


def render_ability_instructions(member: PurchaseGuide) -> list[str]:
    lines = []
    if member.ability_path:
        names = object_dict(member.evidence_summary.get("ability_names")) or {}
        lines.extend([
            "",
            "**Ability order:** "
            + " → ".join(
                str(names.get(str(item), item))
                for item in member.ability_path.ability_ids
            ),
            "Ability evidence: "
            + (
                member.ability_path.fallback_reason
                or (
                    "Build-conditioned observed order."
                    if member.ability_path.filter_item_ids
                    else "Hero-wide observed order."
                )
            ),
        ])
    imbues = {
        item.name: item.imbue_target_ability
        for item in (
            *member.core_purchase_items,
            *(item for items in member.tiers.values() for item in items),
        )
        if item.imbue_target_ability
    }
    lines.extend(f"- Imbue {item}: {ability}." for item, ability in imbues.items())
    return lines

File: crates/deadlock-guides/src/beam_display.rs
use std::fmt::Write as _;

use deadlock_data::{Result, count_as_f64};
use serde_json::Value;

use crate::purchase_guide::PurchaseGuide;

#[must_use]
pub fn generator_metadata(guide: &PurchaseGuide) -> Option<&Value> {
    guide
        .evidence_summary
        .get("generator")
        .filter(|value| value.is_object())
}

#[must_use]
pub fn variant_state_labels(guide: &PurchaseGuide) -> String {
    let rows = variant_state_rows(guide);
    if rows.is_empty() {
        "State unknown".into()
    } else {
        rows.iter()
            .map(|(label, _)| *label)
            .collect::<Vec<_>>()
            .join(", ")
    }
}

fn variant_state_rows(guide: &PurchaseGuide) -> Vec<(&'static str, &Value)> {
    let Some(metadata) = generator_metadata(guide) else {
        return Vec::new();
    };
    let mut rows = Vec::new();
    for (state, label) in [(0, "Behind"), (1, "Even"), (2, "Ahead")] {
        if !metadata["states"]
            .as_array()
            .is_some_and(|states| states.contains(&Value::from(state)))
        {
            continue;
        }
        let row = &metadata["state_evidence"][state.to_string()]["validation"];
        if row["owners"]
            .as_u64()
            .zip(row["wins"].as_u64())
            .is_some_and(|(owners, wins)| owners > 0 && wins <= owners)
        {
            rows.push((label, row));
        }
    }
    rows
}

/// # Errors
/// Returns an error when observation counts exceed the supported numeric range.
pub fn variant_statistics(guide: &PurchaseGuide, detailed: bool) -> Result<Vec<String>> {
    let mut lines = Vec::new();
    for (label, row) in variant_state_rows(guide) {
        let count = deadlock_data::integer(row, "owners")?;
        let wins = deadlock_data::integer(row, "wins")?;
        let rate = 100.0 * count_as_f64(wins)? / count_as_f64(count)?;
        let mut text = format!("{label}: {rate:.1}% | {wins}/{count} wins");
        if detailed {
            write!(
                text,
                " | 95% interval {:.1}%–{:.1}%. State hero baseline: {:.1}% across {} matches.",
                deadlock_data::real(row, "lower_95")? * 100.0,
                deadlock_data::real(row, "upper_95")? * 100.0,
                deadlock_data::real(row, "hero_win_rate")? * 100.0,
                deadlock_data::integer(row, "hero_matches")?
            )?;
            if let Some(cutoff) = row["ownership_before_seconds"]
                .as_u64()
                .filter(|value| *value > 0)
            {
                write!(
                    text,
                    " All final core items owned before {cutoff} seconds; additional items allowed."
                )?;
            }
        }
        lines.push(text);
    }
    Ok(lines)
}

pub fn attach_beam_ability_names(guide: &mut PurchaseGuide, kit: &Value) {
    if generator_metadata(guide).is_none() {
        return;
    }
    let names = kit["abilities"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|row| {
            row["id"]
                .as_u64()
                .zip(row["name"].as_str())
                .map(|(id, name)| (id.to_string(), name.into()))
        })
        .collect();
    guide.evidence_summary["ability_names"] = Value::Object(names);
}
